### PR TITLE
feat: frame aware depth control

### LINKED ISSUE #69: add frame handling for depth control
In `depth_controller_node.py`;
The controller uses the cmd_pose topic to set z depth. But it does not take frame_id into consideration. the frame_id supplied in the message should be the frame of reference. the controller should adjust the base depth, relative to the supplied frame of reference.

### PR DESCRIPTION
**Pull request by eminmeydanoglu**
Closes #69

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -37,3 +37,4 @@
 
 *__pycache__*
 *.pyc
+auv_simulation/
```
Comment (senceryazici) [inline .gitignore:40]: not necessary, or unrelated to this PR

File: auv_control/auv_control/config/default.yaml [hunk 2]
```diff
@@ -23,11 +23,11 @@ model:
     - [0, 0, 0, 0, 0.0, 0]
     - [0, 0, 0, 0, 0, 7.98053]
 
-#kp: [0.0, 0.0, 1.5, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 5.0, 0.0, 1.0]
+#kp: [0.0, 0.0, 1.5, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 0.1, 0.0, 1.0]
```
Comment (senceryazici) [inline auv_control/auv_control/config/default.yaml:26]: these are unrelated, please rollback them

File: auv_control/auv_control/config/default.yaml [hunk 3]
```diff
@@ -23,11 +23,11 @@ model:
...
 #ki: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
 #kd: [0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 0.0, 0.0, 1.0]
 
-kp: [0.0, 0.0, 1.5, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 0.1, 0.0, 1.0]
+kp: [0.0, 0.0, 1.5, 0.0, 0.0, 0.0, 1.0, 1.0, 1.0, 5.0, 0.0, 1.0]
```
Comment (senceryazici) [inline auv_control/auv_control/config/default.yaml:30]: these are unrelated, please rollback them

**Review by senceryazici — CHANGES_REQUESTED** [review]
Good work. thank you for the PR. 

The frame handling and transformations should be on the position controller side (the node that subscribes to cmd_pose.) This requires the cmd_pose type to be geometry_msgs/PoseStamped instead of geometry_msgs/Pose. then the header.frame_id will be interpreted same as your approach.

Adding frame_id to the SetDepth hence to the depth_controller_node is a good idea, with above implementation all you need to do in that node is

```py
def ..callback(req: SetDepthRequest):
    pose_stamped.header.frame_id = req.frame_id
    # continue publishing cmd_pose as usual.
```

**Comment by meydanoglu** [comment]
Hi, made some changes. Looking forward to hear from you @senceryazici

**Review by senceryazici — CHANGES_REQUESTED** [review]
Good work, we're getting there.

File: auv_control/auv_control/scripts/depth_controller_node.py [hunk 4]
```diff
@@ -21,33 +26,51 @@ def __init__(self):
...
 
         # Initialize publisher
-        self.cmd_pose_pub = rospy.Publisher("/taluy/cmd_pose", Pose, queue_size=10)
+        self.cmd_pose_pub = rospy.Publisher(
+            "/taluy/cmd_pose_stamped", PoseStamped, queue_size=10)
 
         # Initialize internal state
         self.target_depth = 0.0
         self.current_depth = 0.0
+        self.target_frame_id = "taluy/base_link" # default to "taluy/base_link" if not set
```
[the remark this answers is not in the document]
Comment (meydanoglu) [reply auv_control/auv_control/scripts/depth_controller_node.py:35]: True, controller_ros.h handles it already

File: auv_control/auv_control/scripts/depth_controller_node.py [hunk 5]
```diff
@@ -3,14 +3,19 @@
...
+import tf2_ros
 
 
 class DepthControllerNode:
     def __init__(self):
         rospy.init_node("depth_controller_node")
+        rospy.loginfo("DepthControllerNode has started.")
+        
+        self.tf_buffer = tf2_ros.Buffer()
+        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer)
```
[the remark this answers is not in the document]
Comment (meydanoglu) [reply auv_control/auv_control/scripts/depth_controller_node.py:18]: Yes, transform handling is now at `controller_ros.h`

File: auv_control/auv_control/scripts/depth_controller_node.py [hunk 6]
```diff
@@ -3,14 +3,19 @@
 import rospy
 from std_msgs.msg import Bool, Float32
 from nav_msgs.msg import Odometry
-from geometry_msgs.msg import Pose
+from geometry_msgs.msg import Pose, PoseStamped
 from auv_msgs.srv import SetDepth, SetDepthRequest, SetDepthResponse
-import time
+import tf2_geometry_msgs
```
[the remark this answers is not in the document]
Comment (meydanoglu) [reply auv_control/auv_control/scripts/depth_controller_node.py:8]: TF libraries are no longer needed here. Also, time was not being used, thus removed as well.

File: auv_control/auv_control/include/auv_control/controller_ros.h [hunk 7]
```diff
@@ -146,12 +150,46 @@ class ControllerROS {
...
+    geometry_msgs::TransformStamped transform_stamped;
+    geometry_msgs::Pose transformed_pose;
+
+    ROS_DEBUG_STREAM_THROTTLE(10.0, "Received cmd_pose with frame_id: " << msg->header.frame_id);
+
+    std::string source_frame = msg->header.frame_id;
+    if (!source_frame.empty() && source_frame[0] == '/') {
+        source_frame = source_frame.substr(1);
+    }
+
```
[the remark this answers is not in the document]
Comment (meydanoglu) [reply auv_control/auv_control/include/auv_control/controller_ros.h:168]: Hmm. In this implementation,
`const auto is_transform_required = !source_frame.empty() || source_frame == base_frame;`
the transformation would be considered required even if source_frame is the same as default frame. But transforming to the same frame seems unnecessary. Do we need to transform the source frame to the default frame if they are the exact same?
This one will not cause unnecessary transformation:
`const auto is_transform_required = !source_frame.empty() && source_frame != default_frame; // only transform if not empty and not the same as default frame`

File: auv_control/auv_control/include/auv_control/controller_ros.h [hunk 8]
```diff
@@ -146,12 +150,46 @@ class ControllerROS {
...
+            transform_stamped = tf_buffer.lookupTransform(base_frame, source_frame, ros::Time(0));
+            tf2::doTransform(msg->pose, transformed_pose, transform_stamped);  
+            ROS_DEBUG_STREAM_THROTTLE(10.0,"Transformed pose from " << source_frame << " to " << base_frame); 
+        } else {
+            ROS_DEBUG_STREAM_THROTTLE(10.0, "Using original pose (no transformation needed)");
+            transformed_pose = msg->pose;
+        }
+    } catch (tf2::TransformException &ex) {
+        ROS_DEBUG_STREAM_THROTTLE(10.0,"Could not transform from " << msg->header.frame_id << " to " << base_frame << ": " << ex.what());
+        transformed_pose = msg->pose;
```
[the remark this answers is not in the document]
Comment (meydanoglu) [reply auv_control/auv_control/include/auv_control/controller_ros.h:180]: Hmm, never thought about that. 
```
catch (tf2::TransformException &ex) { // if transformation failed, ignore the target pose.
        return;
```

File: auv_control/auv_control/scripts/depth_controller_node.py [hunk 9]
```diff
@@ -21,33 +26,51 @@ def __init__(self):
...
         self.alpha = self.dt / (self.tau + self.dt)
-
+        # Set up a timer to call the control_loop method at a rate defined by update_rate
         rospy.Timer(rospy.Duration(1.0 / self.update_rate), self.control_loop)
 
     def target_depth_handler(self, req: SetDepthRequest) -> SetDepthResponse:
+        rospy.loginfo(f"Received SetDepth request with target depth: {req.target_depth} and frame_id: {req.frame_id}")
+    
+        # Check if frame_id is provided
+        if not req.frame_id:
```
[the remark this answers is not in the document]
Comment (meydanoglu) [reply auv_control/auv_control/scripts/depth_controller_node.py:50]: Got it. 
`self.target_frame_id = req.frame_id  # If the target frame_id is not given, simply empty the frame_id`

File: auv_control/auv_control/include/auv_control/controller_ros.h [hunk 10]
```diff
@@ -146,12 +150,46 @@ class ControllerROS {
...
+        source_frame = source_frame.substr(1);
+    }
+
+    try {
+        if (!source_frame.empty() && source_frame != base_frame) {
+            transform_stamped = tf_buffer.lookupTransform(base_frame, source_frame, ros::Time(0));
+            tf2::doTransform(msg->pose, transformed_pose, transform_stamped);  
+            ROS_DEBUG_STREAM_THROTTLE(10.0,"Transformed pose from " << source_frame << " to " << base_frame); 
+        } else {
+            ROS_DEBUG_STREAM_THROTTLE(10.0, "Using original pose (no transformation needed)");
```
[the remark this answers is not in the document]
Comment (meydanoglu) [reply auv_control/auv_control/include/auv_control/controller_ros.h:175]: Changing` base_frame` ---> `default_frame`

File: auv_control/auv_control/include/auv_control/controller_ros.h [hunk 11]
```diff
@@ -146,12 +149,50 @@ class ControllerROS {
...
+
+
+    std::string source_frame = msg->header.frame_id;
+
+    if (!source_frame.empty() && source_frame[0] == '/') {
+      source_frame = source_frame.substr(1);
+    } // the / causes errors.
+    
+    try {
+      const auto is_transform_required =
```
Comment (senceryazici) [inline auv_control/auv_control/include/auv_control/controller_ros.h:169]:
```cpp
const std::optional<std::string> get_source_frame(const std::string& source_frame) {
    if (source_frame.empty()) {
        return std::nullopt;
     } 

    if (source_frame == position_control_default_frame_) {
        return std::nullopt;
    }
    // transform is required.
    if (source_frame[0] == '/') {
        return source_frame.substr(1);
    }
    return source_frame;
}

void callback(.....) {
 // ...
const auto source_frame = get_source_frame(msg->header.frame_id);

auto transformed_pose = msg->pose;
if (source_frame.has_value()) {
    try {
        transformed_pose = ...(from=source_frame.value())...;
        tf2::doTransform....
     } catch ... {
       return;
     }
} 
}

 desired_state_.head(6) = auv::common::conversions::convert<
        geometry_msgs::Pose, ControllerBase::Vector>(transformed_pose);
    latest_command_time_ = ros::Time::now();

```


```

File: auv_control/auv_control/include/auv_control/controller_ros.h [hunk 12]
```diff
@@ -68,13 +71,13 @@ class ControllerROS {
...
     dr_srv_.setCallback(f);
-    
+
     odometry_sub_ =
         nh_.subscribe("odometry", 1, &ControllerROS::odometry_callback, this);
     cmd_vel_sub_ =
         nh_.subscribe("cmd_vel", 1, &ControllerROS::cmd_vel_callback, this);
-    cmd_pose_sub_ =
-        nh_.subscribe("cmd_pose", 1, &ControllerROS::cmd_pose_callback, this);
+    cmd_pose_sub_ = nh_.subscribe("cmd_pose_stamped", 1,
```
[the remark this answers is not in the document]
Comment (meydanoglu) [reply auv_control/auv_control/include/auv_control/controller_ros.h:79]: Alright. Changing the publishing part in depth_controller_node.py

```python
# Initialize publisher
        self.cmd_pose_pub = rospy.Publisher(
            "/taluy/cmd_pose", PoseStamped, queue_size=10)
```

File: auv_control/auv_control/include/auv_control/controller_ros.h [hunk 13]
```diff
@@ -146,10 +151,48 @@ class ControllerROS {
...
+        ROS_DEBUG("Transformed Z value: %f", transformed_pose.position.z);
+      } catch (tf2::TransformException& ex) { // If unsuccessful, return without pose.
+        ROS_DEBUG("Failed to transform pose");
+        return;
+      }
+    }
+
+    desired_state_.head(6) = auv::common::conversions::convert<
+        geometry_msgs::Pose, ControllerBase::Vector>(transformed_pose);
+
```
Comment (meydanoglu) [inline auv_control/auv_control/include/auv_control/controller_ros.h:195]: New cmd_pose_callback structure with a get_source_frame helper function.

File: auv_control/auv_control/include/auv_control/controller_ros.h [hunk 14]
```diff
@@ -264,10 +343,14 @@ void save_parameters() {
...
-  dynamic_reconfigure::Server<auv_control::ControllerConfig> dr_srv_; // Dynamic reconfigure server
-  Eigen::Matrix<double, 12, 1> kp_, ki_, kd_; // Parameters to be dynamically reconfigured
-  std::string config_file_; // Path to the config file
+  dynamic_reconfigure::Server<auv_control::ControllerConfig>
+      dr_srv_;  // Dynamic reconfigure server
+  Eigen::Matrix<double, 12, 1> kp_, ki_,
+      kd_;                   // Parameters to be dynamically reconfigured
+  std::string config_file_;  // Path to the config file
+
+  std::string position_control_default_frame_;
```
Comment (meydanoglu) [inline auv_control/auv_control/include/auv_control/controller_ros.h:352]: member variable decleration for ` position_control_default_frame_`

File: auv_control/auv_control/scripts/depth_controller_node.py [hunk 15]
```diff
@@ -21,33 +21,43 @@ def __init__(self):
         )
 
         # Initialize publisher
-        self.cmd_pose_pub = rospy.Publisher("/taluy/cmd_pose", Pose, queue_size=10)
+        self.cmd_pose_pub = rospy.Publisher(
+            "/taluy/cmd_pose", PoseStamped, queue_size=10)
```
Comment (meydanoglu) [inline auv_control/auv_control/scripts/depth_controller_node.py:25]: publishes to `cmd_pose` instead of `cmd_pose_stamped`

**Review by meydanoglu — COMMENTED** [review]
Sure. Here is a video of me testing the new code:



https://github.com/user-attachments/assets/a3874903-e829-42d1-8090-d3a19f63f67f

> What is happening here?
1. I call the service without any tf, and the code fails to transform (nothing passed to `desired_state_`)
2. I define tf's and watch the `transformed_pose` change.
3. I kill the tf, and function now fails to transform again. (since there is no valid transform now.)


### Fix for a potential bug
We used to use `ros::Time(0)` for the transformation time. But this might lead to some unwanted problems, for example;

"_Maybe there existed a transformation between `frame_1` and `default_frame` 10 minutes ago. We received a service call with a `target_pose` in `frame_1` (maybe not likely, but still). Even though there's currently no transformation between these frames and we are absolutely not sure the last known transform is still valid, our code would have gone ahead and transformed the pose anyway. Now, that won't happen thanks to `ros::Time::now()` using the time of the function callback._"

**Comment by senceryazici** [comment]
Looks good, I'm curious about the hardware tests though.

**Comment by meydanoglu** [comment]
> Looks good, I'm curious about the hardware tests though.
We will be testing the features of the new implementation in the next pool test.

**Comment by eminmeydanoglu** [comment]
In the last pool test, we successfully validated the frame-aware depth control functionality. Its only that my frame aware controller had an interesting interaction with @talhabw 's heading control implementation. Since my code was transforming everything in command pose, it also transformed the last three components, causing the heading control to request alignment with depth control's reference frame. So, I modified the transform handling process to only transform the z axis pose. The depth control now correctly responds to frame-specific commands while not causing heading control to go silly.


Irrelevantly, I want to mention that the way we handle command timeout is not bug-free. In this PR, notice how we avoid updating the desired state if there is a problem
```cpp
void cmd_pose_callback(const geometry_msgs::PoseStamped::ConstPtr& msg) {
         ..... 
         try {
         //transform z
      } catch (tf2::TransformException& ex) { // If unsuccessful, exit and don't update desired state
        ROS_DEBUG("Failed to transform pose");
        return;

```
So if the source frame is missing, we dont update the desired state (which is correct). But guess what, the desired state will stay the same with the last update unless there comes a new service call. So if the frame is gone missing after it was used even once, that desired state update will be kept forever. For more information about the issue refer to https://github.com/itu-auv/auv-software/issues/115

**Comment by eminmeydanoglu** [comment]
This PR was thoroughly tested during the last pool test, including edge cases and its interaction with heading control, and everything worked flawlessly. I believe it’s ready to go! @senceryazici

**Review by senceryazici — DISMISSED** [review]

**Comment by eminmeydanoglu** [comment]
Hey @senceryazici, thanks for approving the PR. I want to share some thoughts on the current design, as I find it not so much cool:
The structure is,
depth_controller_node -> publish z pose value in reference frame -> controller_ros.h -> transform to odom frame -> write to desired_state matrix.  

1. Confusing cmd_pose values: This means the values in the cmd_pose message are not (always) in odom frame. cmd_pose.position.z = -0.5 might not mean that robot wants to go down 0.5 meters from water level--maybe something entirely different once transformed. This can be confusing and misleading for someone checking the cmd_pose commands. And it’s not immediately clear what the actual pose.z in the odom frame is (or should we print the desired state matrix :D)
2. SoC: Seems to me like processing the desired depth is a more of a "depth controller node" job. After all, it is responsible for deciding where we want to go depth-wise. I think controller_ros.h shouldn't be responsible of that? 

So I think it is more convenient to move the transformation part back to depth controller node. I am curious about your thoughts!

**Review by ozanhakantunca — APPROVED** [review]
